hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java

```diff
@@ -932,12 +934,36 @@ public NativeQueryImplementor createNativeQuery(String sqlString, Class resultCl
 	}
 
 	private void handleNativeQueryResult(NativeQueryImplementor query, Class resultClass) {
+		boolean isObjectArray = Object[].class.equals( resultClass );
+
 		if ( Tuple.class.equals( resultClass ) ) {
 			query.setResultTransformer( new NativeQueryTupleTransformer() );
 		}
-		else {
+		else if ( resultClass.isArray() && !isObjectArray ) {
+			Class elementClass = resultClass.getComponentType();
+
+			query.setResultTransformer( new BasicTransformerAdapter() {
+				@Override
+				public Object transformTuple(Object[] tuple, String[] aliases) {
+					Object[] result = (Object[]) Array.newInstance( elementClass, tuple.length );
+					for ( int i = 0; i < tuple.length; i++ ) {
+						result[i] = elementClass.cast( tuple[i] );
+					}
+					return result;
+				}
+			} );
+		}
+		else if ( this.getFactory().getMetamodel().entityPersisters().get( resultClass.getName() ) != null ) {
```

Comment:
Maybe I'm wrong but I think the key of the map returned by the `entityPersisters()` is not always the class name but can also be the entity name if defined (at least if the mapping is done via XML, it seems that there is a possibility the class name is not used).

See https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/hbm/AbstractEntitySourceImpl.java#L114

That being said, it looks like `MetamodelImpl#entityPersister(Class entityClass)` is doing the same mistake so I'm not sure about it.